### DISCUSSION TITLE
How to convert cube-map MTLTexture as filament Texture for IBL?

### DISCUSSION THREAD

**Discussion by Ankbzpx** [google/filament#4721] [Q&A] [answered]
The goal is to use ARkit estimated [environmentTexture](https://developer.apple.com/documentation/arkit/arenvironmentprobeanchor/2977511-environmenttexture) for realtime IBL. 

I attempted to convert it as filament Texture by using [import](https://github.com/google/filament/blob/main/filament/include/filament/Texture.h#L238) with format `backend::TextureFormat::RGBA16F` and sampler `backend::SamplerType::SAMPLER_CUBEMAP`. It was then prefiltered by [iblprefilter](https://github.com/google/filament/tree/main/libs/iblprefilter) as the reflection for building new IBL. However, it seemed that I missed something as the model was completely unlit (only black silhouette).

What would be the correct way of converting it?

`DebugDescription` for `environmentTexture`
```
<CaptureMTLTexture: 0x2807298c0> -> <AGXA14FamilyTexture: 0x1231a5250>
    label = com.apple.arkit.gpusphericalblur.target 
    textureType = MTLTextureTypeCube 
    pixelFormat = MTLPixelFormatRGBA16Float 
    width = 256 
    height = 256 
    depth = 1 
    arrayLength = 1 
    mipmapLevelCount = 9 
    sampleCount = 1 
    cpuCacheMode = MTLCPUCacheModeDefaultCache 
    storageMode = MTLStorageModeShared 
    hazardTrackingMode = MTLHazardTrackingModeTracked 
    resourceOptions = MTLResourceCPUCacheModeDefaultCache MTLResourceStorageModeShared MTLResourceHazardTrackingModeTracked  
    usage = MTLTextureUsageShaderRead MTLTextureUsageShaderWrite MTLTextureUsageRenderTarget MTLTextureUsagePixelFormatView 
    shareable = 0 
    framebufferOnly = 0 
    purgeableState = MTLPurgeableStateNonVolatile 
    swizzle = [MTLTextureSwizzleRed, MTLTextureSwizzleGreen, MTLTextureSwizzleBlue, MTLTextureSwizzleAlpha] 
    isCompressed = 0 
    parentTexture = <null> 
    parentRelativeLevel = 0 
    parentRelativeSlice = 0 
    buffer = <null> 
    bufferOffset = 0 
    bufferBytesPerRow = 0 
    iosurface = 0x0 
    iosurfacePlane = 0 
    allowGPUOptimizedContents = YES
    label = com.apple.arkit.gpusphericalblur.target
```

**Comment by romainguy** [comment, marked as answer]
@bejado

**Reply by bejado** [reply]
@Ankbzpx In Xcode under "Edit Scheme -> Diagnostics", can you make sure Metal "API Validation" and "Shader Validation" are enabled? It would also be helpful to see how you're calling into `iblprefilter`.

**Reply by Ankbzpx** [reply]
I toggled "API Validation" and "Shader Validation" as suggested but it didn't seem to catch any useful info
```
[Metal Diagnostics Warning] Application Deployment Target Version (14.1) does not match OS Version (14.8) - diagnostics may be missing debug information

2021-10-13 11:09:43.186956+0800 APP[14350:5557617] Metal GPU Frame Capture Enabled
2021-10-13 11:09:43.207688+0800 APP[14350:5557617] Metal API Validation Enabled
2021-10-13 11:09:43.208291+0800 APP[14350:5557617] Metal GPU Validation Enabled

FEngine (64 bits) created at 0x11eb40000 (threading is enabled)
FEngine resolved backend: Metal
Selected physical device 'Apple A12Z GPU'
Supported GPU families: 
  MTLGPUFamilyCommon3
  MTLGPUFamilyApple5
```
I am afraid I can't do anything with mismatch warning as Xcode doesn't provide `iOS 14.8` Application Deployment Target Version option (It only provides `iOS 9.0` till `iOS 14.7` and `iOS 15.0`)


My code was adapted from [hello-ar example](https://github.com/google/filament/tree/main/ios/samples/hello-ar)
```
    // App-specific
    struct {
        ...
	IndirectLight* indirectLight = nullptr;
        Texture* iblTexture = nullptr;
        IBLPrefilterContext* context = nullptr;
        IBLPrefilterContext::SpecularFilter* filter = nullptr;
    } app;

```
```
void FilamentApp::setupFilament() {
    ...
    app.context = new IBLPrefilterContext(*engine);
    app.filter = new IBLPrefilterContext::SpecularFilter(*app.context);
}
```
```
void FilamentApp::updateIBL(uint32_t width, uint32_t height, uint32_t depth, uint32_t levels, intptr_t texture) {
    scene->setIndirectLight(nullptr);
    engine->destroy(app.indirectLight);
    engine->destroy(app.iblTexture);
    
    app.iblTexture = (*app.filter)(Texture::Builder()
                                   .width(width)
                                   .height(height)
                                   .depth(depth)
                                   .levels(levels)
                                   .import(texture)
                                   .format(backend::TextureFormat::RGBA16F)
                                   .sampler(backend::SamplerType::SAMPLER_CUBEMAP)
                                   .build(*engine));
    
    app.indirectLight = IndirectLight::Builder()
        .reflections(app.iblTexture)
        .intensity(30000)
        .build(*engine);
    scene->setIndirectLight(app.indirectLight);
}
```
`updateIBL` was called in OC wrapper once per second
```
- (void)updateAmbient:(id<MTLTexture>)environmentTexture 
{
    app->updateIBL((uint32_t)environmentTexture.width,
                   (uint32_t)environmentTexture.height,
                   (uint32_t)environmentTexture.depth,
                   (uint32_t)environmentTexture.mipmapLevelCount,
                   (intptr_t) CFBridgingRetain(environmentTexture));
}
```

**Reply by Ankbzpx** [reply]
On second thought, I caught following error in another test device
```
FEngine (64 bits) created at 0x160500000 (threading is enabled)
FEngine resolved backend: Metal
Selected physical device 'Apple A12 GPU'
Supported GPU families: 
  MTLGPUFamilyCommon3
  MTLGPUFamilyApple5

<MTLBufferArgument: 0x282194e40>
        Name = vertexBuffer.0 
        Type = MTLArgumentTypeBuffer 
        Access = MTLArgumentAccessReadOnly 
        LocationIndex = 0 
        IsActive = 1 
        ArrayLength = 1 
        TypeInfo = 
            DataType = MTLDataTypePointer 
            ElementType = MTLDataTypeStruct  
            Access = MTLArgumentAccessReadOnly 
            Alignment = 4 
            DataSize = 32
        Alignment = 4 
        DataSize = 32 
        DataType = MTLDataTypeStruct
buffer: "<unknown>"
 __:::__ __delegate_identifier__: GPUToolsDiagnostics
2021-10-13 11:40:46.156145+0800 APP[71418:20655997] [GPUDebug] Invalid device load executing vertex function "tessellation_vertex_quad" encoder: "Render Command Encoder", draw: 1, at offset 18446744073709549832

<MTLBufferArgument: 0x282194e40>
        Name = vertexBuffer.0 
        Type = MTLArgumentTypeBuffer 
        Access = MTLArgumentAccessReadOnly 
        LocationIndex = 0 
        IsActive = 1 
        ArrayLength = 1 
        TypeInfo = 
            DataType = MTLDataTypePointer 
            ElementType = MTLDataTypeStruct  
            Access = MTLArgumentAccessReadOnly 
            Alignment = 4 
            DataSize = 32
        Alignment = 4 
        DataSize = 32 
        DataType = MTLDataTypeStruct
buffer: "<unknown>"
2021-10-13 11:40:46.166448+0800 APP[71418:20655990] [Metal Diagnostics] __message__: Invalid device load executing vertex function "tessellation_vertex_quad" encoder: "Render Command Encoder", draw: 1, at offset 18446744073709549832
```

**Reply by bejado** [reply]
I was able to import the `environmentTexture` into Filament and set it as the IBL texture. Make sure your model's material is using a low roughness value so you can see the reflections. I'm not sure what those error messages mean, Filament doesn't perform any tessellation.

As a side note, it's probably unnecessary to go through our `iblprefilter` library. The environment texture already contains a mipmap chain, and I suspect ARKit is performing its own filtering.

**Reply by bejado** [reply]
Something else: [Apple's `environmentTexture` documentation](https://developer.apple.com/documentation/arkit/arenvironmentprobeanchor/2977511-environmenttexture?language=objc) seems to think that the texture format will be `MTLPixelFormatBGRA8Unorm_sRGB`, when in actuality it's `MTLPixelFormatRGBA16Float`.

**Reply by Ankbzpx** [reply]
Thanks for your help! I am very grateful for your patience!

It turned out I made a mistake by manually placing `AREnvironmentProbeAnchor` at the place of `ARPlaneAnchor`, which isn't where it should be placed. When I leave it to auto, IBL works as expected.

A side question, with prefiltered iblTexture, it is still possible to retrieve sh for direction estimate?

**Reply by ThomasGorisse** [reply]
Nice to see that someone is doing it on iOS (and facing the same issues/questions I faced 2 months ago when doing it on Android)
I think that this PR will be a mine of gold for what you are currently trying to achieve: https://github.com/ThomasGorisse/SceneformMaintained/pull/156
Feel free to ask me questions about making the two worlds living together (I have also spend a lot of time on RealityKit and have some knowledge on how ARKit works)

Do you have a public repo?

**Reply by Ankbzpx** [reply]
Appreciate your sharing and helpfulness! I am working at a company so unfortunately the repo is proprietary that I am not allowed to share. Thanks anyway!

**Reply by Ankbzpx** [reply]
For future reference, [`environmentTexture`](https://developer.apple.com/documentation/arkit/arenvironmentprobeanchor/2977511-environmenttexture) in `AREnvironmentProbeAnchor` can be imported to filament Cubemap Texture by:

In OC:
```
    updateIBLTexture((uint32_t)environmentTexture.width,
                           (uint32_t)environmentTexture.height,
                           (uint32_t)environmentTexture.depth,
                           (uint32_t)environmentTexture.mipmapLevelCount,
                           (intptr_t) CFBridgingRetain(environmentTexture)
                    );
```
In CPP:
```
void updateIBLTexture(uint32_t width, uint32_t height, uint32_t depth, uint32_t levels, intptr_t texture)
    iblTexture = Texture::Builder()
        .width(width)
        .height(height)
        .depth(depth)
        .levels(levels)
        .import(texture)
        .format(backend::TextureFormat::RGBA16F)
        .sampler(backend::SamplerType::SAMPLER_CUBEMAP)
        .build(*engine);
```

As suggested by @bejado, `environmentTexture` was already prefiltered so that it can be directly used for `reflections` when building IBL without going through `iblprefilter` library.

To achieve smooth transition between change of `environmentTexture`, a [metal kernel](https://developer.apple.com/documentation/metal/processing_a_texture_in_a_compute_function) can be used to `mix` two `iblTextures` to generate a new one. Don't forget to `generateMipmaps` for newly mixed `iblTexture` before passing it to filament.
